File: ai/zen/zen.mjs
#!/usr/bin/env node
// ZEN Node.js ES Module Wrapper
import { createRequire } from 'module';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { spawn, execSync } from 'child_process';
import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'fs';
import sqlite3 from 'sqlite3';
import { open } from 'sqlite';
import crypto from 'crypto';

const __dirname = dirname(fileURLToPath(import.meta.url));
const require = createRequire(import.meta.url);

class ZenMemory {
    constructor(dbPath) {
        this.dbPath = dbPath;
        this.db = null;
    }

    async init() {
        this.db = await open({
            filename: this.dbPath,
            driver: sqlite3.Database
        });

        await this.db.exec(`
            CREATE TABLE IF NOT EXISTS memories (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                hash TEXT UNIQUE,
                content TEXT,
                embedding BLOB,
                agent_id INTEGER,
                phase REAL,
                entropy REAL,
                tokens INTEGER,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                accessed_at DATETIME DEFAULT CURRENT_TIMESTAMP
            );
            
            CREATE TABLE IF NOT EXISTS clusters (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                centroid BLOB,
                size INTEGER,
                variance REAL,
                last_updated DATETIME DEFAULT CURRENT_TIMESTAMP
            );
            
            CREATE TABLE IF NOT EXISTS tokens (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                session TEXT,
                agent_id INTEGER,
                token TEXT,
                position INTEGER,
                entropy REAL,
                timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
            );
            
            CREATE INDEX IF NOT EXISTS idx_memories_hash ON memories(hash);
            CREATE INDEX IF NOT EXISTS idx_memories_phase ON memories(phase);
            CREATE INDEX IF NOT EXISTS idx_tokens_session ON tokens(session);
        `);
    }

    async storeMemory(content, agentId = 0, phase = 0.0, entropy = 0.5) {
        const hash = crypto.createHash('sha256').update(content).digest('hex');
        const embedding = await this.generateEmbedding(content);
        
        await this.db.run(
            `INSERT OR REPLACE INTO memories 
             (hash, content, embedding, agent_id, phase, entropy, accessed_at)
             VALUES (?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)`,
            [hash, content, embedding, agentId, phase, entropy]
        );
        
        return hash;
    }

    async generateEmbedding(text) {
        // Use Ollama embeddings API
        const response = await fetch('http://localhost:11434/api/embeddings', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                model: 'nomic-embed-text',
                prompt: text
            })
        });
        
        const data = await response.json();
        return Buffer.from(data.embedding);
    }

    async recallSemantic(query, limit = 5) {
        const queryEmbedding = await this.generateEmbedding(query);
        
        return await this.db.all(`
            SELECT content, 
                   agent_id,
                   phase,
                   entropy,
                   dot_product(embedding, ?) as similarity
            FROM memories
            ORDER BY similarity DESC
            LIMIT ?
        `, [queryEmbedding, limit]);
    }
}

class ZenAgent {
    constructor(id, name, phase) {
        this.id = id;
        this.name = name;
        this.phase = phase;
        this.entropy = 0.5;
    }

    async process(input, context = {}) {
        const temperature = 0.7 + (this.entropy * 0.3);
        const streamPath = join(__dirname, 'streams', 
            `agent_${this.id}_${Date.now()}.txt`);

        const response = await this.callOllama(input, temperature, streamPath);
        
        // Update entropy based on response characteristics
        this.entropy = this.calculateEntropy(response);
        
        return {
            agent: this.name,
            phase: this.phase,
            response: response,
            entropy: this.entropy,
            stream: streamPath
        };
    }

    calculateEntropy(text) {
        // Calculate Shannon entropy of text
        const freq = {};
        for (const char of text) {
            freq[char] = (freq[char] || 0) + 1;
        }
        
        let entropy = 0;
        const len = text.length;
        for (const count of Object.values(freq)) {
            const p = count / len;
            entropy -= p * Math.log2(p);
        }
        
        return Math.min(entropy / 8, 0.9); // Normalize
    }

    async callOllama(prompt, temperature, streamPath) {
        return new Promise((resolve, reject) => {
            const proc = spawn('ollama', [
                'run', 
                'llama3.2:3b',
                prompt
            ], {
                env: { ...process.env, OLLAMA_HOST: 'http://localhost:11434' }
            });

            let output = '';
            const stream = fs.createWriteStream(streamPath);

            proc.stdout.on('data', (data) => {
                const text = data.toString();
                output += text;
                stream.write(text);
            });

            proc.stderr.on('data', (data) => {
                console.error(`Agent ${this.name} error:`, data.toString());
            });

            proc.on('close', (code) => {
                stream.end();
                resolve(output);
            });

            proc.on('error', reject);
        });
    }
}

export { ZenMemory, ZenAgent };

File: ai/zen/zen.py
#!/usr/bin/env python3
"""
ZEN Python Wrapper for ML/Embedding Operations
"""
import sqlite3
import numpy as np
import hashlib
import json
import sys
import os
from pathlib import Path
from typing import List, Tuple, Optional
import requests
from datetime import datetime

class ZenEmbedder:
    """Handle text embeddings and similarity"""
    
    def __init__(self, ollama_host="http://localhost:11434"):
        self.ollama_host = ollama_host
        self.cache = {}
        
    def embed_text(self, text: str, model: str = "nomic-embed-text") -> np.ndarray:
        """Get text embedding from Ollama"""
        cache_key = hashlib.sha256(text.encode()).hexdigest()
        
        if cache_key in self.cache:
            return self.cache[cache_key]
        
        try:
            response = requests.post(
                f"{self.ollama_host}/api/embeddings",
                json={"model": model, "prompt": text},
                timeout=30
            )
            response.raise_for_status()
            embedding = np.array(response.json()["embedding"], dtype=np.float32)
            self.cache[cache_key] = embedding
            return embedding
        except Exception as e:
            print(f"Embedding error: {e}", file=sys.stderr)
            # Fallback to simple TF-IDF like embedding
            return self.fallback_embed(text)
    
    def fallback_embed(self, text: str) -> np.ndarray:
        """Simple fallback embedding"""
        words = text.lower().split()
        vocab = list(set(words))
        embedding = np.zeros(384, dtype=np.float32)
        
        for i, word in enumerate(vocab[:384]):
            embedding[i] = words.count(word) / len(words)
        
        norm = np.linalg.norm(embedding)
        if norm > 0:
            embedding = embedding / norm
        
        return embedding
    
    def cosine_similarity(self, a: np.ndarray, b: np.ndarray) -> float:
        """Calculate cosine similarity"""
        dot = np.dot(a, b)
        norm_a = np.linalg.norm(a)
        norm_b = np.linalg.norm(b)
        
        if norm_a == 0 or norm_b == 0:
            return 0.0
        
        return float(dot / (norm_a * norm_b))
    
    def cluster_embeddings(self, embeddings: List[np.ndarray], k: int = 8) -> List[List[int]]:
        """Simple K-means clustering"""
        from sklearn.cluster import KMeans
        
        if len(embeddings) < k:
            k = len(embeddings)
        
        if k <= 1:
            return [list(range(len(embeddings)))]
        
        kmeans = KMeans(n_clusters=k, random_state=42, n_init=10)
        clusters = kmeans.fit_predict(embeddings)
        
        result = [[] for _ in range(k)]
        for idx, cluster_id in enumerate(clusters):
            result[cluster_id].append(idx)
        
        return result

class ZenEntropy:
    """Entropy and chaos management for parallel reasoning"""
    
    def __init__(self, seed: int = None):
        self.seed = seed or int(datetime.now().timestamp() * 1000)
        np.random.seed(self.seed)
        self.entropy_pool = []
        self.decay_rate = 0.95
        
    def generate_entropy(self, n: int = 8) -> List[float]:
        """Generate entropy values for agents"""
        entropies = np.random.uniform(0.1, 0.9, n)
        
        # Apply 2π/8 phase distribution
        phases = np.linspace(0, 2 * np.pi, n, endpoint=False)
        phase_mod = (np.sin(phases) + 1) / 2  # Convert to 0-1 range
        
        entropies = entropies * 0.7 + phase_mod * 0.3
        return entropies.tolist()
    
    def adjust_entropy(self, current: float, feedback: float) -> float:
        """Adjust entropy based on feedback"""
        new_entropy = current * self.decay_rate + feedback * (1 - self.decay_rate)
        return np.clip(new_entropy, 0.1, 0.9)
    
    def calculate_chaos_index(self, responses: List[str]) -> float:
        """Calculate chaos index from multiple responses"""
        if not responses:
            return 0.0
        
        # Measure diversity between responses
        similarities = []
        embedder = ZenEmbedder()
        
        for i in range(len(responses)):
            for j in range(i + 1, len(responses)):
                emb_i = embedder.embed_text(responses[i])
                emb_j = embedder.embed_text(responses[j])
                sim = embedder.cosine_similarity(emb_i, emb_j)
                similarities.append(sim)
        
        if similarities:
            avg_similarity = np.mean(similarities)
            chaos = 1.0 - avg_similarity
        else:
            chaos = 0.5
        
        return float(chaos)

class ZenParser:
    """Parse various input types"""
    
    @staticmethod
    def parse_input(input_str: str) -> dict:
        """Parse prompt/url/file/hash/walletseed"""
        result = {
            "type": "prompt",
            "content": input_str,
            "valid": True
        }
        
        # URL detection
        if input_str.startswith(("http://", "https://", "ftp://")):
            result["type"] = "url"
            try:
                response = requests.get(input_str, timeout=10)
                result["content"] = response.text[:10000]  # Limit
            except:
                result["valid"] = False
        
        # File detection
        elif os.path.exists(input_str):
            result["type"] = "file"
            try:
                with open(input_str, 'r', encoding='utf-8') as f:
                    result["content"] = f.read(10000)  # Limit
            except:
                result["valid"] = False
        
        # Hash detection (64 hex chars)
        elif len(input_str) == 64 and all(c in "0123456789abcdefABCDEF" for c in input_str):
            result["type"] = "hash"
        
        # Wallet seed detection (12 or 24 words)
        elif 12 <= len(input_str.split()) <= 24:
            result["type"] = "walletseed"
        
        return result
    
    @staticmethod
    def batch_process(file_path: str, operation: str = "crud") -> List[dict]:
        """Batch process file with CRUD/SOAP/REST operations"""
        results = []
        
        if not os.path.exists(file_path):
            return results
        
        with open(file_path, 'r') as f:
            for line_num, line in enumerate(f, 1):
                line = line.strip()
                if not line:
                    continue
                
                result = {
                    "line": line_num,
                    "input": line,
                    "operation": operation,
                    "status": "pending"
                }
                
                try:
                    if operation == "crud":
                        # CRUD operations on local data
                        parsed = ZenParser.parse_input(line)
                        result["parsed"] = parsed
                        result["status"] = "success"
                    
                    elif operation == "soap":
                        # SOAP-like operation
                        result["status"] = "soap_not_implemented"
                    
                    elif operation == "rest":
                        # REST API call
                        if line.startswith("http"):
                            response = requests.get(line, timeout=5)
                            result["response_code"] = response.status_code
                            result["status"] = "success"
                    
                    results.append(result)
                    
                except Exception as e:
                    result["status"] = f"error: {str(e)}"
                    results.append(result)
        
        return results

if __name__ == "__main__":
    # Command line interface
    if len(sys.argv) > 1:
        parser = ZenParser()
        
        if sys.argv[1] == "parse":
            result = parser.parse_input(sys.argv[2])
            print(json.dumps(result, indent=2))
        
        elif sys.argv[1] == "embed":
            embedder = ZenEmbedder()
            embedding = embedder.embed_text(sys.argv[2])
            print(f"Embedding shape: {embedding.shape}")
            print(f"First 10 values: {embedding[:10]}")
        
        elif sys.argv[1] == "entropy":
            entropy = ZenEntropy()
            values = entropy.generate_entropy(8)
            print(f"Entropy values: {values}")
            print(f"Chaos index: {entropy.calculate_chaos_index(sys.argv[2:])}")
